source/MRMesh/MRPointsComponents.cpp

```diff
@@ -53,7 +53,61 @@ Expected<MR::VertBitSet> getLargestComponentsUnion( const PointCloud& pointCloud
     return result;
 }
 
-Expected<UnionFind<MR::VertId>> getUnionFindStructureVerts( const PointCloud& pointCloud, float maxDist, const VertBitSet* region /*= nullptr*/, ProgressCallback pc /*= {}*/ )
+Expected<std::vector<VertBitSet>> getLargestComponents( const PointCloud& pointCloud, float maxDist, int minSize, ProgressCallback pc /*= {} */ )
+{
+    MR_TIMER
+
+    assert( maxDist > 0.f );
+    assert( minSize > 1 );
+    const auto& validPoints = pointCloud.validPoints;
+    ProgressCallback subPc = subprogress( pc, 0.f, 0.9f );
+    auto unionStructsRes = getUnionFindStructureVerts( pointCloud, maxDist, nullptr, subPc );
+    if ( !unionStructsRes.has_value() )
+        return unexpectedOperationCanceled();
+    auto& unionStructs = *unionStructsRes;
+    const auto& allRoots = unionStructs.roots();
+
+    subPc = subprogress( pc, 0.9f, 0.95f );
+    int counter = 0;
+    const float counterMax = float( validPoints.count() );
+    const int counterDivider = int( validPoints.count() ) / 100;
+    HashMap<VertId, int> root2size;
+    for ( auto v : validPoints )
+    {
+        ++root2size[allRoots[v]];
+        if ( !reportProgress( subPc, counter / counterMax, counter, counterDivider ) )
+            return unexpectedOperationCanceled();
+    }
+
+    subPc = subprogress( pc, 0.95f, 1.f );
+    counter = 0;
+    std::vector<VertBitSet> result;
+    HashMap<VertId, int> root2index;
+    const size_t validPointsSize = validPoints.find_last() + 1;
+    for ( auto v : validPoints )
+    {
+        const VertId root = allRoots[v];
+        if ( root2size[root] >= minSize )
+        {
+            int index = -1;
+            if ( root2index.contains( root ) )
+                index = root2index[root];
+            else
+            {
+                index = int( result.size() );
+                root2index[root] = index;
+                result.push_back( VertBitSet( validPointsSize ) );
+            }
```

Comment:
instead of two calls to hash-map: `contains` and `[]` better use only one:
```
auto [it, inserted] = root2index.insert( { root, result.size() } );
if ( inserted )
...

Reply:
fixed